File: src/inference/derive-beliefs.ts
// Derive PublicBeliefs from accumulated FactConstraintIR[].
// Replaces the old mergeInferences() approach with a lossless constraint-first model.

import type { Seat, Suit } from "../engine/types";
import { Suit as SuitEnum } from "../engine/types";
import type { FactConstraintIR } from "../core/contracts/agreement-module";
import type { PublicBeliefs, DerivedRanges, QualitativeConstraint } from "../core/contracts/inference";

const ALL_SUITS: Suit[] = [SuitEnum.Spades, SuitEnum.Hearts, SuitEnum.Diamonds, SuitEnum.Clubs];

/** Map factId suffix → engine Suit value for suit-length constraints. */
const SUIT_SUFFIX: Record<string, Suit> = {
  spades: SuitEnum.Spades,
  hearts: SuitEnum.Hearts,
  diamonds: SuitEnum.Diamonds,
  clubs: SuitEnum.Clubs,
};

/** Known qualitative factIds and their human-readable labels. */
const QUALITATIVE_LABELS: Record<string, string> = {
  "bridge.hasFourCardMajor": "Has 4-card major",
  "bridge.hasFiveCardMajor": "Has 5-card major",
  "bridge.hasShortage": "Has shortage (0-1 in a suit)",
};

/**
 * Derive PublicBeliefs from accumulated constraints for a single seat.
 * Source of truth is always the raw constraints array — ranges and
 * qualitative labels are computed from it.
 */
export function derivePublicBeliefs(
  seat: Seat,
  constraints: readonly FactConstraintIR[],
): PublicBeliefs {
  return {
    seat,
    constraints,
    ranges: deriveRanges(constraints),
    qualitative: deriveQualitative(constraints),
  };
}

/**
 * Compute flat display-friendly ranges from constraints.
 * Handles: hand.hcp, hand.suitLength.*, hand.isBalanced.
 */
function deriveRanges(constraints: readonly FactConstraintIR[]): DerivedRanges {
  let hcpMin = 0;
  let hcpMax = 40;
  let balanced: boolean | undefined;

  const suitMins: Record<string, number> = {
    [SuitEnum.Spades]: 0,
    [SuitEnum.Hearts]: 0,
    [SuitEnum.Diamonds]: 0,
    [SuitEnum.Clubs]: 0,
  };
  const suitMaxes: Record<string, number> = {
    [SuitEnum.Spades]: 13,
    [SuitEnum.Hearts]: 13,
    [SuitEnum.Diamonds]: 13,
    [SuitEnum.Clubs]: 13,
  };

  for (const c of constraints) {
    // HCP
    if (c.factId === "hand.hcp") {
      if (c.operator === "gte" && typeof c.value === "number") {
        hcpMin = Math.max(hcpMin, c.value);
      } else if (c.operator === "lte" && typeof c.value === "number") {
        hcpMax = Math.min(hcpMax, c.value);
      } else if (c.operator === "range" && typeof c.value === "object" && c.value !== null && "min" in c.value) {
        const range = c.value as { min: number; max: number };
        hcpMin = Math.max(hcpMin, range.min);
        hcpMax = Math.min(hcpMax, range.max);
      }
      continue;
    }

    // Suit lengths
    const suitMatch = c.factId.match(/^hand\.suitLength\.(\w+)$/);
    if (suitMatch) {
      const suit = SUIT_SUFFIX[suitMatch[1]!];
      if (!suit) continue;
      if (c.operator === "gte" && typeof c.value === "number") {
        suitMins[suit] = Math.max(suitMins[suit]!, c.value);
      } else if (c.operator === "lte" && typeof c.value === "number") {
        suitMaxes[suit] = Math.min(suitMaxes[suit]!, c.value);
      }
      continue;
    }

    // Balanced
    if (c.factId === "hand.isBalanced" && c.operator === "boolean") {
      balanced = c.value === true;
    }
  }

  // Balanced distributions (4-3-3-3, 4-4-3-2, 5-3-3-2) guarantee ≥2 in every suit
  if (balanced === true) {
    for (const suit of ALL_SUITS) {
      suitMins[suit] = Math.max(suitMins[suit]!, 2);
    }
  }

  // Clamp contradictions
  if (hcpMin > hcpMax) hcpMin = hcpMax;

  const suitLengths: Record<Suit, { min: number; max: number }> = {
    [SuitEnum.Spades]: { min: 0, max: 13 },
    [SuitEnum.Hearts]: { min: 0, max: 13 },
    [SuitEnum.Diamonds]: { min: 0, max: 13 },
    [SuitEnum.Clubs]: { min: 0, max: 13 },
  };
  for (const suit of ALL_SUITS) {
    let min = suitMins[suit]!;
    const max = suitMaxes[suit]!;
    if (min > max) min = max;
    suitLengths[suit] = { min, max };
  }

  return {
    hcp: { min: hcpMin, max: hcpMax },
    suitLengths,
    isBalanced: balanced,
  };
}

/**
 * Extract qualitative constraints — those that don't reduce to flat per-suit ranges.
 * These are displayed as-is in the UI (e.g. "Has 4-card major").
 */
function deriveQualitative(constraints: readonly FactConstraintIR[]): QualitativeConstraint[] {
  const result: QualitativeConstraint[] = [];

  for (const c of constraints) {
    const label = QUALITATIVE_LABELS[c.factId];
    if (label) {
      // Only show positive assertions for now
      if (c.operator === "boolean" && c.value === true) {
        // Deduplicate
        if (!result.some((q) => q.factId === c.factId)) {
          result.push({ factId: c.factId, label, operator: c.operator, value: c.value });
        }
      }
    }
  }

  return result;
}

/**
 * Convert a HandInference to FactConstraintIR[].
 * Used at the boundary where InferenceProvider returns HandInference
 * but the public beliefs pipeline needs FactConstraintIR[].
 */
export function handInferenceToConstraints(inf: {
  readonly minHcp?: number;
  readonly maxHcp?: number;
  readonly isBalanced?: boolean;
  readonly suits: Partial<Record<Suit, { readonly minLength?: number; readonly maxLength?: number }>>;
}): FactConstraintIR[] {
  const constraints: FactConstraintIR[] = [];

  if (inf.minHcp !== undefined) {
    constraints.push({ factId: "hand.hcp", operator: "gte", value: inf.minHcp });
  }
  if (inf.maxHcp !== undefined) {
    constraints.push({ factId: "hand.hcp", operator: "lte", value: inf.maxHcp });
  }
  if (inf.isBalanced === true) {
    constraints.push({ factId: "hand.isBalanced", operator: "boolean", value: true });
  }

  for (const [suit, si] of Object.entries(inf.suits)) {
    if (!si) continue;
    const suffix = suitToSuffix(suit as Suit);
    if (!suffix) continue;
    if (si.minLength !== undefined) {
      constraints.push({ factId: `hand.suitLength.${suffix}`, operator: "gte", value: si.minLength });
    }
    if (si.maxLength !== undefined) {
      constraints.push({ factId: `hand.suitLength.${suffix}`, operator: "lte", value: si.maxLength });
    }
  }

  return constraints;
}

/** Map Suit enum value to factId suffix. */
function suitToSuffix(suit: Suit): string | null {
  switch (suit) {
    case SuitEnum.Spades: return "spades";
    case SuitEnum.Hearts: return "hearts";
    case SuitEnum.Diamonds: return "diamonds";
    case SuitEnum.Clubs: return "clubs";
    default: return null;
  }
}

File: src-tauri/crates/bridge-session/src/inference/derive_beliefs.rs
//! Derive `PublicBeliefs` from accumulated `FactConstraint[]`.
//!
//! Constraint-first model: accumulated constraints per seat are the source of truth.
//! `derive_public_beliefs()` computes `DerivedRanges` (HCP min/max, per-suit length
//! min/max) and qualitative constraints from the raw constraint array.

use bridge_conventions::types::meaning::{ConstraintValue, FactConstraint, FactOperator};
use bridge_engine::types::{Seat, Suit};
use std::collections::HashMap;

use super::types::{
    DerivedRanges, HandInference, NumberRange, PublicBeliefs, QualitativeConstraint,
};

const ALL_SUITS: [Suit; 4] = [Suit::Spades, Suit::Hearts, Suit::Diamonds, Suit::Clubs];

/// Map factId suffix to engine Suit value for suit-length constraints.
fn suit_from_suffix(suffix: &str) -> Option<Suit> {
    match suffix {
        "spades" => Some(Suit::Spades),
        "hearts" => Some(Suit::Hearts),
        "diamonds" => Some(Suit::Diamonds),
        "clubs" => Some(Suit::Clubs),
        _ => None,
    }
}

/// Map Suit to factId suffix.
fn suit_to_suffix(suit: Suit) -> &'static str {
    match suit {
        Suit::Spades => "spades",
        Suit::Hearts => "hearts",
        Suit::Diamonds => "diamonds",
        Suit::Clubs => "clubs",
    }
}

/// Known qualitative factIds and their human-readable labels.
fn qualitative_label(fact_id: &str) -> Option<&'static str> {
    match fact_id {
        "bridge.hasFourCardMajor" => Some("Has 4-card major"),
        "bridge.hasFiveCardMajor" => Some("Has 5-card major"),
        "bridge.hasShortage" => Some("Has shortage (0-1 in a suit)"),
        _ => None,
    }
}

/// Extract a u32 from a `ConstraintValue`, handling Number variant.
fn value_as_u32(v: &ConstraintValue) -> Option<u32> {
    match v {
        ConstraintValue::Number(n) => n.as_u64().map(|n| n as u32),
        _ => None,
    }
}

/// Extract min/max from a Range `ConstraintValue`.
fn value_as_range(v: &ConstraintValue) -> Option<(u32, u32)> {
    match v {
        ConstraintValue::Range { min, max } => {
            let min = min.as_u64()? as u32;
            let max = max.as_u64()? as u32;
            Some((min, max))
        }
        _ => None,
    }
}

/// Derive `PublicBeliefs` from accumulated constraints for a single seat.
/// Source of truth is always the raw constraints array -- ranges and
/// qualitative labels are computed from it.
pub fn derive_public_beliefs(seat: Seat, constraints: &[FactConstraint]) -> PublicBeliefs {
    PublicBeliefs {
        seat,
        constraints: constraints.to_vec(),
        ranges: derive_ranges(constraints),
        qualitative: derive_qualitative(constraints),
    }
}

/// Compute flat display-friendly ranges from constraints.
/// Handles: hand.hcp, hand.suitLength.*, hand.isBalanced.
fn derive_ranges(constraints: &[FactConstraint]) -> DerivedRanges {
    let mut hcp_min: u32 = 0;
    let mut hcp_max: u32 = 40;
    let mut balanced: Option<bool> = None;

    let mut suit_mins: HashMap<Suit, u32> = ALL_SUITS.iter().map(|&s| (s, 0u32)).collect();
    let mut suit_maxes: HashMap<Suit, u32> = ALL_SUITS.iter().map(|&s| (s, 13u32)).collect();

    for c in constraints {
        // HCP constraints
        if c.fact_id == "hand.hcp" {
            match c.operator {
                FactOperator::Gte => {
                    if let Some(v) = value_as_u32(&c.value) {
                        hcp_min = hcp_min.max(v);
                    }
                }
                FactOperator::Lte => {
                    if let Some(v) = value_as_u32(&c.value) {
                        hcp_max = hcp_max.min(v);
                    }
                }
                FactOperator::Range => {
                    if let Some((min, max)) = value_as_range(&c.value) {
                        hcp_min = hcp_min.max(min);
                        hcp_max = hcp_max.min(max);
                    }
                }
                _ => {}
            }
            continue;
        }

        // Suit length constraints: hand.suitLength.<suit>
        if let Some(suffix) = c.fact_id.strip_prefix("hand.suitLength.") {
            if let Some(suit) = suit_from_suffix(suffix) {
                match c.operator {
                    FactOperator::Gte => {
                        if let Some(v) = value_as_u32(&c.value) {
                            *suit_mins.get_mut(&suit).unwrap() = suit_mins[&suit].max(v);
                        }
                    }
                    FactOperator::Lte => {
                        if let Some(v) = value_as_u32(&c.value) {
                            *suit_maxes.get_mut(&suit).unwrap() = suit_maxes[&suit].min(v);
                        }
                    }
                    _ => {}
                }
            }
            continue;
        }

        // Balanced constraint
        if c.fact_id == "hand.isBalanced" && c.operator == FactOperator::Boolean {
            if let ConstraintValue::Bool(b) = c.value {
                balanced = Some(b);
            }
        }
    }

    // Balanced distributions (4-3-3-3, 4-4-3-2, 5-3-3-2) guarantee >= 2 in every suit
    if balanced == Some(true) {
        for suit in &ALL_SUITS {
            let entry = suit_mins.get_mut(suit).unwrap();
            *entry = (*entry).max(2);
        }
    }

    // Clamp contradictions
    if hcp_min > hcp_max {
        hcp_min = hcp_max;
    }

    let mut suit_lengths = HashMap::new();
    for suit in &ALL_SUITS {
        let mut min = suit_mins[suit];
        let max = suit_maxes[suit];
        if min > max {
            min = max;
        }
        suit_lengths.insert(*suit, NumberRange { min, max });
    }

    DerivedRanges {
        hcp: NumberRange {
            min: hcp_min,
            max: hcp_max,
        },
        suit_lengths,
        is_balanced: balanced,
    }
}

/// Extract qualitative constraints -- those that don't reduce to flat per-suit ranges.
fn derive_qualitative(constraints: &[FactConstraint]) -> Vec<QualitativeConstraint> {
    let mut result = Vec::new();

    for c in constraints {
        if let Some(label) = qualitative_label(&c.fact_id) {
            // Only show positive assertions for now
            if c.operator == FactOperator::Boolean {
                if let ConstraintValue::Bool(true) = c.value {
                    // Deduplicate
                    if !result
                        .iter()
                        .any(|q: &QualitativeConstraint| q.fact_id == c.fact_id)
                    {
                        result.push(QualitativeConstraint {
                            fact_id: c.fact_id.clone(),
                            label: label.to_string(),
                            operator: "boolean".to_string(),
                            value: serde_json::Value::Bool(true),
                        });
                    }
                }
            }
        }
    }

    result
}

/// Convert a `HandInference` to `Vec<FactConstraint>`.
/// Used at the boundary where `InferenceProvider` returns `HandInference`
/// but the public beliefs pipeline needs `FactConstraint[]`.
pub fn hand_inference_to_constraints(inf: &HandInference) -> Vec<FactConstraint> {
    let mut constraints = Vec::new();

    if let Some(min) = inf.min_hcp {
        constraints.push(FactConstraint {
            fact_id: "hand.hcp".to_string(),
            operator: FactOperator::Gte,
            value: ConstraintValue::int(min as i64),
            is_public: None,
        });
    }
    if let Some(max) = inf.max_hcp {
        constraints.push(FactConstraint {
            fact_id: "hand.hcp".to_string(),
            operator: FactOperator::Lte,
            value: ConstraintValue::int(max as i64),
            is_public: None,
        });
    }
    if inf.is_balanced == Some(true) {
        constraints.push(FactConstraint {
            fact_id: "hand.isBalanced".to_string(),
            operator: FactOperator::Boolean,
            value: ConstraintValue::Bool(true),
            is_public: None,
        });
    }

    for (suit, si) in &inf.suits {
        let suffix = suit_to_suffix(*suit);
        if let Some(min) = si.min_length {
            constraints.push(FactConstraint {
                fact_id: format!("hand.suitLength.{}", suffix),
                operator: FactOperator::Gte,
                value: ConstraintValue::int(min as i64),
                is_public: None,
            });
        }
        if let Some(max) = si.max_length {
            constraints.push(FactConstraint {
                fact_id: format!("hand.suitLength.{}", suffix),
                operator: FactOperator::Lte,
                value: ConstraintValue::int(max as i64),
                is_public: None,
            });
        }
    }

    constraints
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_constraints_produce_maximally_loose_beliefs() {
        let beliefs = derive_public_beliefs(Seat::North, &[]);
        assert_eq!(beliefs.ranges.hcp, NumberRange { min: 0, max: 40 });
        assert_eq!(beliefs.ranges.is_balanced, None);
        for suit in &ALL_SUITS {
            assert_eq!(
                beliefs.ranges.suit_lengths[suit],
                NumberRange { min: 0, max: 13 }
            );
        }
        assert!(beliefs.qualitative.is_empty());
    }

    #[test]
    fn hcp_gte_narrows_min() {
        let constraints = vec![FactConstraint {
            fact_id: "hand.hcp".to_string(),
            operator: FactOperator::Gte,
            value: ConstraintValue::int(15),
            is_public: None,
        }];
        let beliefs = derive_public_beliefs(Seat::South, &constraints);
        assert_eq!(beliefs.ranges.hcp.min, 15);
        assert_eq!(beliefs.ranges.hcp.max, 40);
    }

    #[test]
    fn hcp_lte_narrows_max() {
        let constraints = vec![FactConstraint {
            fact_id: "hand.hcp".to_string(),
            operator: FactOperator::Lte,
            value: ConstraintValue::int(17),
            is_public: None,
        }];
        let beliefs = derive_public_beliefs(Seat::South, &constraints);
        assert_eq!(beliefs.ranges.hcp.min, 0);
        assert_eq!(beliefs.ranges.hcp.max, 17);
    }

    #[test]
    fn hcp_range_narrows_both() {
        let constraints = vec![FactConstraint {
            fact_id: "hand.hcp".to_string(),
            operator: FactOperator::Range,
            value: ConstraintValue::Range {
                min: serde_json::Number::from(15),
                max: serde_json::Number::from(17),
            },
            is_public: None,
        }];
        let beliefs = derive_public_beliefs(Seat::South, &constraints);
        assert_eq!(beliefs.ranges.hcp, NumberRange { min: 15, max: 17 });
    }

    #[test]
    fn suit_length_constraints() {
        let constraints = vec![
            FactConstraint {
                fact_id: "hand.suitLength.spades".to_string(),
                operator: FactOperator::Gte,
                value: ConstraintValue::int(5),
                is_public: None,
            },
            FactConstraint {
                fact_id: "hand.suitLength.clubs".to_string(),
                operator: FactOperator::Lte,
                value: ConstraintValue::int(3),
                is_public: None,
            },
        ];
        let beliefs = derive_public_beliefs(Seat::South, &constraints);
        assert_eq!(beliefs.ranges.suit_lengths[&Suit::Spades].min, 5);
        assert_eq!(beliefs.ranges.suit_lengths[&Suit::Clubs].max, 3);
    }

    #[test]
    fn balanced_constraint_sets_min_suit_length_2() {
        let constraints = vec![FactConstraint {
            fact_id: "hand.isBalanced".to_string(),
            operator: FactOperator::Boolean,
            value: ConstraintValue::Bool(true),
            is_public: None,
        }];
        let beliefs = derive_public_beliefs(Seat::South, &constraints);
        assert_eq!(beliefs.ranges.is_balanced, Some(true));
        for suit in &ALL_SUITS {
            assert_eq!(beliefs.ranges.suit_lengths[suit].min, 2);
        }
    }

    #[test]
    fn contradiction_clamping() {
        let constraints = vec![
            FactConstraint {
                fact_id: "hand.hcp".to_string(),
                operator: FactOperator::Gte,
                value: ConstraintValue::int(20),
                is_public: None,
            },
            FactConstraint {
                fact_id: "hand.hcp".to_string(),
                operator: FactOperator::Lte,
                value: ConstraintValue::int(10),
                is_public: None,
            },
        ];
        let beliefs = derive_public_beliefs(Seat::South, &constraints);
        // Clamp: min should equal max
        assert_eq!(beliefs.ranges.hcp.min, beliefs.ranges.hcp.max);
    }

    #[test]
    fn qualitative_constraints() {
        let constraints = vec![FactConstraint {
            fact_id: "bridge.hasFourCardMajor".to_string(),
            operator: FactOperator::Boolean,
            value: ConstraintValue::Bool(true),
            is_public: None,
        }];
        let beliefs = derive_public_beliefs(Seat::South, &constraints);
        assert_eq!(beliefs.qualitative.len(), 1);
        assert_eq!(beliefs.qualitative[0].label, "Has 4-card major");
    }

    #[test]
    fn qualitative_deduplication() {
        let constraints = vec![
            FactConstraint {
                fact_id: "bridge.hasFourCardMajor".to_string(),
                operator: FactOperator::Boolean,
                value: ConstraintValue::Bool(true),
                is_public: None,
            },
            FactConstraint {
                fact_id: "bridge.hasFourCardMajor".to_string(),
                operator: FactOperator::Boolean,
                value: ConstraintValue::Bool(true),
                is_public: None,
            },
        ];
        let beliefs = derive_public_beliefs(Seat::South, &constraints);
        assert_eq!(beliefs.qualitative.len(), 1);
    }

    #[test]
    fn hand_inference_to_constraints_roundtrip() {
        let inf = HandInference {
            seat: Seat::North,
            min_hcp: Some(15),
            max_hcp: Some(17),
            is_balanced: Some(true),
            suits: {
                let mut m = HashMap::new();
                m.insert(
                    Suit::Spades,
                    super::super::types::SuitInference {
                        min_length: Some(5),
                        max_length: None,
                    },
                );
                m
            },
            source: "test".to_string(),
        };
        let constraints = hand_inference_to_constraints(&inf);
        // Should have: hcp gte 15, hcp lte 17, isBalanced true, suitLength.spades gte 5
        assert_eq!(constraints.len(), 4);

        let beliefs = derive_public_beliefs(Seat::North, &constraints);
        assert_eq!(beliefs.ranges.hcp, NumberRange { min: 15, max: 17 });
        assert_eq!(beliefs.ranges.is_balanced, Some(true));
        assert_eq!(beliefs.ranges.suit_lengths[&Suit::Spades].min, 5);
    }
}
